lib/broadway/rate_limiter.ex

```diff
@@ -0,0 +1,68 @@
+defmodule Broadway.RateLimiter do
+  @moduledoc false
+
+  use GenServer
+  @row_name :rate_limit_counter
+
+  def start_link(opts) do
+    case Keyword.fetch!(opts, :rate_limiting) do
+      # If we don't have rate limiting options, we don't even need to start this rate
+      # limiter process.
+      nil ->
+        :ignore
+
+      rate_limiting_opts ->
+        name = Keyword.fetch!(opts, :name)
+        genserver_opts = [name: Module.concat(name, __MODULE__)]
+        GenServer.start_link(__MODULE__, {name, rate_limiting_opts}, genserver_opts)
+    end
+  end
+
+  def rate_limit(table_name, amount)
+      when is_atom(table_name) and is_integer(amount) and amount > 0 do
+    :ets.update_counter(table_name, @row_name, -amount)
+  end
+
+  def get_currently_allowed(table_name) when is_atom(table_name) do
+    :ets.lookup_element(table_name, @row_name, 2)
+  end
+
+  def table_name(broadway_name) when is_atom(broadway_name) do
+    Module.concat(broadway_name, RateLimiterETS)
+  end
+
+  @impl true
+  def init({name, rate_limiting_opts}) do
+    interval = Keyword.fetch!(rate_limiting_opts, :interval)
+    allowed = Keyword.fetch!(rate_limiting_opts, :allowed_messages)
+
+    table_name = table_name(name)
+
+    _ets = :ets.new(table_name, [:named_table, :public, :set])
+    :ets.insert(table_name, {@row_name, allowed})
+
+    _ = schedule_next_reset(interval, allowed)
```

Comment:
Make `allowed` part of the state instead of passing it in the message because if we want configurable rate limittng in the future, then it should be in the state. :) Also, even if you don't want to "future plan", all other values are on the state, so let's keep it consistent!

Reply:
Or maybe you are doing it as part of the message for testing? If so, ignore this comment for now.